src/Tag/Table.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Html\Tag;
+
+use Yiisoft\Html\Tag\Base\NormalTag;
+
+/**
+ * @link https://www.w3.org/TR/html52/tabular-data.html#the-table-element
+ */
+final class Table extends NormalTag
+{
+    private ?Caption $caption = null;
+    private array $columnGroups = [];
+    private array $columns = [];
+    private ?Thead $header = null;
+    private array $body = [];
+    private array $rows = [];
+    private ?Tfoot $footer = null;
+
+    public function caption(?Caption $caption): self
+    {
+        $new = clone $this;
+        $new->caption = $caption;
+        return $new;
+    }
+
+    public function captionString(string $content, bool $encode = true): self
+    {
+        $caption = Caption::tag()->content($content);
+        if (!$encode) {
+            $caption = $caption->encode(false);
+        }
+        return $this->caption($caption);
+    }
+
+    /**
+     * @param Colgroup ...$columnGroups One or more column groups ({@see Colgroup}).
+     */
+    public function columnGroups(Colgroup ...$columnGroups): self
+    {
+        $new = clone $this;
+        $new->columnGroups = $columnGroups;
+        return $new;
+    }
+
+    /**
+     * @param Colgroup ...$columnGroups One or more column groups ({@see Colgroup}).
+     */
+    public function addColumnGroups(Colgroup ...$columnGroups): self
+    {
+        $new = clone $this;
+        $new->columnGroups = [...$new->columnGroups, ...$columnGroups];
+        return $new;
+    }
+
+    /**
+     * @param Col ...$columns One or more columns ({@see Col}).
+     */
+    public function columns(Col ...$columns): self
+    {
+        $new = clone $this;
+        $new->columns = $columns;
+        return $new;
+    }
+
+    /**
+     * @param Col ...$columns One or more columns ({@see Col}).
+     */
+    public function addColumns(Col ...$columns): self
+    {
+        $new = clone $this;
+        $new->columns = [...$new->columns, ...$columns];
```

Comment:
Can you bench this vs
```suggestion
        foreach ($columns as $column) {
            $new->columns[] = $column;
        }
```
on different sets?

Reply:
Benchmark:

```php
/**
 * @Iterations(50)
 */
final class ArrayBench
{
    private array $columns = ['A', 'B', 'C'];
    private array $add = ['D', 'E', 'F', 'G', 'H', 'J', 'K', 'L'];

    public function benchA()
    {
        $columns = [...$this->columns, ...$this->add];
    }

    public function benchB()
    {
        $columns = $this->columns;
        foreach ($this->add as $column) {
            $columns[] = $column;
        }
    }
}
```

Result:

```
λ vendor\bin\phpbench.bat run tests\ArrayBench.php --report=aggregate
PhpBench @git_tag@. Running benchmarks.
Using configuration file: F:\open-source\yiisoft\html/phpbench.json

\Yiisoft\Html\Tests\ArrayBench

    benchA..................................I49 [μ Mo]/r: 11.800 11.068 (μs) [μSD μRSD]/r: 2.315μs 19.62%
    benchB..................................I49 [μ Mo]/r: 12.240 11.703 (μs) [μSD μRSD]/r: 2.818μs 23.02%

2 subjects, 100 iterations, 2 revs, 0 rejects, 0 failures, 0 warnings
(best [mean mode] worst) = 10.000 [12.020 11.386] 24.000 (μs)
⅀T: 1,202.000μs μSD/r 2.567μs μRSD/r: 21.322%
suite: 134628e58bbe3de58bda86e8d30537570c812977, date: 2021-03-18, stime: 20:27:07
+------------+---------+-----+------+-----+------------+----------+----------+----------+----------+---------+--------+-------+
| benchmark  | subject | set | revs | its | mem_peak   | best     | mean     | mode     | worst    | stdev   | rstdev | diff  |
+------------+---------+-----+------+-----+------------+----------+----------+----------+----------+---------+--------+-------+
| ArrayBench | benchA  | 0   | 1    | 50  | 4,656,544b | 10.000μs | 11.800μs | 11.068μs | 24.000μs | 2.315μs | 19.62% | 1.00x |
| ArrayBench | benchB  | 0   | 1    | 50  | 4,656,544b | 10.000μs | 12.240μs | 11.703μs | 25.000μs | 2.818μs | 23.02% | 1.04x |
+------------+---------+-----+------+-----+------------+----------+----------+----------+----------+---------+--------+-------+
```

Approximately the same.

Reply:
Run line
`./vendor/bin/phpbench run tests/benchmark/ArrayBench.php --report='extends: "aggregate", break: ["benchmark", "revs", "groups"], sort: {set: "asc"}' --progress=dots`
Benchmark
```php
<?php

/**
 * @Iterations(1)
 * @Revs(5000)
 */
final class ArrayBench
{
    public function firstData()
    {
        return [
            'empty-3' => [[], $this->getData(3)],
            'empty-50' => [[], $this->getData(50)],
        ];
    }

    public function secondData()
    {
        return [
            '3-3' => [$this->getData(3), $this->getData(3)],
            '3-50' => [$this->getData(3), $this->getData(50)],
        ];
    }

    public function thirdData()
    {
        return [
            '50-3' => [$this->getData(50), $this->getData(3)],
            '50-50' => [$this->getData(50), $this->getData(50)],
        ];
    }

    private function getData(int $count): array
    {
        $data = [];
        for ($i = 1; $i <= $count; $i++) {
            $data[] = (string)$i;
        }
        return $data;
    }

    /**
     * @Groups("first")
     * @ParamProviders("firstData")
     */
    public function benchAwithFirstData(array $params): void
    {
        $columns = [...$params[0], ...$params[1]];
    }

    /**
     * @Groups("first")
     * @ParamProviders("firstData")
     */
    public function benchBwithFirstData(array $params): void
    {
        $columns = $params[0];
        foreach ($params[1] as $column) {
            $columns[] = $column;
        }
    }

    /**
     * @Groups("second")
     * @ParamProviders("secondData")
     */
    public function benchAwithSecondData(array $params): void
    {
        $columns = [...$params[0], ...$params[1]];
    }

    /**
     * @Groups("second")
     * @ParamProviders("secondData")
     */
    public function benchBwithSecondData(array $params): void
    {
        $columns = $params[0];
        foreach ($params[1] as $column) {
            $columns[] = $column;
        }
    }

    /**
     * @Groups("third")
     * @ParamProviders("thirdData")
     */
    public function benchAwithThirdData(array $params): void
    {
        $columns = [...$params[0], ...$params[1]];
    }

    /**
     * @Groups("third")
     * @ParamProviders("thirdData")
     */
    public function benchBwithThirdData(array $params): void
    {
        $columns = $params[0];
        foreach ($params[1] as $column) {
            $columns[] = $column;
        }
    }
}

```

Reply:
Create repo with benchmarks: https://github.com/vjik/php-unpacking-vs-foreach-bench